planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -0,0 +1,438 @@
+            shortestDistance(other.to, current.to, current.from, loopMinLength - other.length);
+          if (distance <= loopMinLength) {
+            loops.add(new HasLoop(other, distance));
+          }
+        }
+        if (loops.size() > 1) {
+          HasLoop min = loops.stream().min(Comparator.comparingDouble(HasLoop::distance)).get();
+          for (var loop : loops) {
+            if (loop != min) {
+              loop.edge.remove();
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private double shortestDistance(Node start, Node end, Node exclude, double maxLength) {
+    Map<Integer, Double> bestDistance = new HashMap<>();
+    record Candidate(Node node, double cost, double heuristic) {}
+    PriorityQueue<Candidate> frontier = new PriorityQueue<>(Comparator.comparingDouble(Candidate::heuristic));
+    if (exclude != start) {
+      frontier.offer(new Candidate(start, 0, start.distance(end)));
+    }
+    while (!frontier.isEmpty()) {
+      Candidate candidate = frontier.poll();
+      Node current = candidate.node;
+      if (current == end) {
+        return candidate.cost;
+      }
+
+      for (var edge : current.getEdges()) {
+        var neighbor = edge.to;
+        if (neighbor != exclude) {
+          double newDist = candidate.cost + edge.length;
+          double prev = bestDistance.getOrDefault(neighbor.id, Double.POSITIVE_INFINITY);
+          if (newDist < prev) {
+            bestDistance.put(neighbor.id, newDist);
+            double heuristic = newDist + neighbor.distance(end);
+            if (heuristic <= maxLength) {
+              frontier.offer(new Candidate(neighbor, newDist, heuristic));
+            }
+          }
+        }
+      }
+    }
+    return Double.POSITIVE_INFINITY;
+  }
+
+  private void removeShortStubEdges(double stubMinLength) {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        if (edge.length < stubMinLength &&
+          (edge.from.getEdges().size() == 1 || edge.to.getEdges().size() == 1 || edge.from == edge.to)) {
+          edge.remove();
+        }
+      }
+    }
+  }
+
+  private void removeShortEdges() {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        if (edge.length < minLength) {
+          edge.remove();
+        }
+      }
+    }
+  }
+
+  public List<LineString> getMergedLineStrings() {
+    List<List<Coordinate>> edges = nodeLines(input);
+    buildNodes(edges);
+
+    merge();
+
+    if (loopMinLength > 0.0) {
+      breakLoops();
+      merge();
+    }
+
+    if (minLength > 0.0) {
+      double step = 1.0 / precisionModel.getScale();
+      for (double stubMinLength = 0.0; stubMinLength < minLength; stubMinLength += step) {
+        removeShortStubEdges(stubMinLength);
+        merge();
+      }
+      removeShortStubEdges(minLength);
+      merge();
+      // minLength = 10 * 0.0625;
+      // removeShortEdges();
+      // merge();
```

Comment:
OK great! I can add a `simplifyEdges()` method

Reply:
oops I see you added that already! Updated it to use a specialized version of `DouglasPeuckerSimplifier` that works on `List<Coordinate>` to avoid the extra transformation.

Reply:
I'm not sure about the dropping stubs step, at least for transportation_name layer... I tried the latest version on Massachusetts and it's looking like we have shorter line segments than old JTS version

<table>
<tr>
<td>
<img width="559" alt="image" src="https://github.com/user-attachments/assets/10bc7a7a-785f-4839-b760-a6be1848800d">
</td>
<td>
<img width="562" alt="image" src="https://github.com/user-attachments/assets/9858a86a-ca84-4b78-9351-f65af2991901">

</td>
</tr>
</table>

But if I remove the drop stubs step then JTS and loop liner merger output look pretty close


<table>
<tr>
<td>
<img width="559" alt="image" src="https://github.com/user-attachments/assets/10bc7a7a-785f-4839-b760-a6be1848800d">
</td>
<td>
<img width="551" alt="image" src="https://github.com/user-attachments/assets/ae1126ab-f2b0-4c5e-943d-5edf1a262cea">


</td>
</tr>
</table>